### PR TITLE
Persistence supporting multiple relationships

### PR DESCRIPTION
**Pull request by mrsimpson**
addresses #15 
see https://github.com/graywolf336/temporary-rocketlets-ts-definition/pull/10#discussion_r124967308

I'm struggling a bit to get persistence properly integrated. For the above reasons, I'd like to have it as part of the signatures still.

WDYT?

### CODE REVIEW

File: src/Rocketlet.ts [hunk 1]
```diff
@@ -83,8 +84,7 @@ export abstract class Rocketlet {
      * extended by this Rocketlet have one or multiple Rocketlet data items?
      *
      */
-    // TODO: Discuss what the purpose of this is, commenting out for now.
-    // public abstract getRocketChatAssociation(): IRocketChatAssociation;
+    public abstract getRocketChatAssociations(): Array<IRocketChatAssociation>;
```
Comment (mrsimpson) [inline src/Rocketlet.ts]: I sind't see how implementing different interfaces would help here, so I just enhanced this metadata-retrieval to support - literally - multiple associations

File: src/accessors/IPersistence.ts [hunk 2]
```diff
@@ -1,28 +1,10 @@
-import { IPersistenceRead } from './IPersistenceRead';
```
Comment (mrsimpson) [inline src/accessors/IPersistence.ts]: This interface provides a wrapper for persistence access

File: src/accessors/IPersistence.ts [hunk 3]
```diff
@@ -1,28 +1,10 @@
...
-     * Updates
-     * @param id The Rocketlet data's ID (from create)
-     */
-    update(id: string, data: object): void; // todo: might be better off with IExecutionResult returned
+export interface IPersistence {
 
-    remove(id: string);
-
-    removeByReference(referenceId: string);
+    relatedTo(model: RocketChatAssociationModel): IRelatedPersistence;
```
Comment (mrsimpson) [inline src/accessors/IPersistence.ts]: consumption:
```javascript
persistence.relatedTo(RocketChatAssociationModel.ROOM).create(roomId, myRocketletData);
```

feels fine for me.
`relatedTo` could raise an exception for contract violation if the model is not exposed as relatedAssociation

**Comment by graywolf336** [comment]
I can see where this would be useful, however I'm concerned that by adding `IRocketChatAssociation` and requiring it we are starting to make it extremely complicated to get started...

**Comment by mrsimpson** [comment]
> however I'm concerned that by adding IRocketChatAssociation and requiring it we are starting to make it extremely complicated to get started...
Actually, the interface is not needed for the persistence access. If the consumer passes the model to be related to anyway, the association information actually only served the purpose of transparency.

**Comment by graywolf336** [comment]
Makes sense and I realized it wasn't 100% required but would act as a transparency, I just was worried about how "complicated" it would end up making getting users started writing Rocketlets who haven't been following it this whole time.

**Comment by mrsimpson** [comment]
@graywolf336 I believe it to be very generic, in fact.  

> the example that comes to mind is a todo-list Rocketlet that creates items on the list in the room via a new slash command

This is exactly one use case which I would tackle with this `relatedPersistence`. As the todo-items relate to the room,  you could create a todo-item like that:

`const todoItemId = persistence.relatedTo(RocketChatAssociationModel.ROOM).create(this.roomId, todoItem);`

**Comment by graywolf336** [comment]
Heh you're right and I even thought about that when writing the example but it didn't fully dawn on me. :+1:

**Comment by mrsimpson** [comment]
the only thing which isn't covered yet is if some plugin is completely detached from Rocket.Chat, say a text templates Rocketlet which allows you to select quick-texts (without relating to rooms or users). Then, a customPersistence might make sense.
On the other hand, this would not be stored within the RocketChat DB then at all: It should be stored in the application's persistence totally.

**Pull request merged**